tasks/connectors/edgescan/lib/kenna_api.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module Kenna
+  module Toolkit
+    module Edgescan
+      class KennaApi
+        include Kenna::Toolkit::KdiHelpers
+
+        def initialize(options)
+          @kenna_api_host = options[:kenna_api_host]
+          @kenna_api_key = options[:kenna_api_key]
+          @kenna_connector_id = options[:kenna_connector_id]
+          @output_dir = "#{$basedir}/#{options[:output_directory]}"
+        end
+
+        # Converts an Edgescan asset into Kenna friendly ones and adds them into memory
+        #
+        # Note: Edgescan and Kenna assets don't map one to one. A Kenna asset is more like an
+        #       Edgescan location specifier. Because of that, one Edgescan asset usually gets turned
+        #       into multiple Kenna assets.
+        def add_assets(edgescan_asset, existing_kenna_assets)
+          edgescan_asset.to_kenna_assets(existing_kenna_assets).each do |kenna_asset|
+            add_asset(kenna_asset)
+          end
+        end
+
+        # Converts Edgescan vulnerabilities into Kenna ones and adds them into memory
+        def add_vulnerabilities(edgescan_vulnerabilities)
+          edgescan_vulnerabilities.each do |vulnerability|
+            add_vulnerability(vulnerability.external_asset_id, vulnerability.to_kenna_vulnerability)
+          end
+        end
+
+        # Converts Edgescan definitions into Kenna ones and adds them into memory
+        def add_definitions(edgescan_definitions)
+          edgescan_definitions.each do |edgescan_definition|
+            add_definition(edgescan_definition.to_kenna_definition)
+          end
+        end
+
+        # Uploads whatever's in memory into Kenna and then clears memory
+        #
+        # Note: Uploaded data does not get imported into Kenna automatically. It just sits there
+        #       until `kickoff` is called.
+        #       This allows for uploading in batches. Once a few batches have been uploaded and
+        #       you're happy for whatever is there to get imported into Kenna you can call `kickoff`
+        def upload
+          kdi_upload(@output_dir, "batch-#{millis}.json", @kenna_connector_id, @kenna_api_host, @kenna_api_key)
+        end
+
+        # Kicks off connector tasks so that whatever was uploaded actually gets imported into Kenna
+        def kickoff
+          kdi_connector_kickoff(@kenna_connector_id, @kenna_api_host, @kenna_api_key)
+        end
+
+        # Fetches existing assets tagged with these Edgescan IDs from the Kenna API
+        def fetch_assets_with_edgescan_ids(edgescan_ids)
```

Comment:
let's talk about why this is needed. most of the toolkit connectors don't pull assets from Kenna. Kenna will automatically match up assets in an inbound KDI file to existing assets in Kenna.